twill-yarn/src/test/java/org/apache/twill/yarn/LogLevelChangeTestRun.java

```diff
@@ -0,0 +1,249 @@
+          i++;
+        }
+
+        try {
+          TimeUnit.MILLISECONDS.sleep(100);
+        } catch (InterruptedException e) {
+          break;
+        }
+      }
+    }
+
+    @Override
+    public void stop() {
+      if (runThread != null) {
+        runThread.interrupt();
+      }
+    }
+  }
+
+  /**
+   * Second runnable.
+   */
+  public static final class LogLevelTestSecondRunnable extends AbstractTwillRunnable {
+    public static final Logger LOG = LoggerFactory.getLogger(LogLevelChangeTestRun.LogLevelTestSecondRunnable.class);
+
+    private volatile Thread runThread;
+
+    @Override
+    public void run() {
+      this.runThread = Thread.currentThread();
+
+      // check if the initial log level is DEBUG
+      Assert.assertTrue(LOG.isDebugEnabled() && !LOG.isTraceEnabled());
+
+      int i = 0;
+      while (!Thread.interrupted()) {
+        if (i == 0 && !LOG.isDebugEnabled()) {
+          // check if the log level is changed to INFO
+          Assert.assertTrue(LOG.isInfoEnabled());
+          i++;
+        }
+        if (i == 1 && LOG.isDebugEnabled()) {
+          // check if the log level is changed to TRACE
+          Assert.assertTrue(LOG.isTraceEnabled());
+          i++;
+        }
+
+        try {
+          TimeUnit.MILLISECONDS.sleep(100);
+        } catch (InterruptedException e) {
+          break;
+        }
+      }
+    }
+
+    @Override
+    public void stop() {
+      if (runThread != null) {
+        runThread.interrupt();
+      }
+    }
+  }
+
+  /**
+   * A test TwillApplication to test setting log level to DEBUG.
+   */
+  public static final class LogLevelTestApplication implements TwillApplication {
+
+    @Override
+    public TwillSpecification configure() {
+      return TwillSpecification.Builder.with()
+        .setName("LogLevelChangeTest")
+        .withRunnable()
+        .add(LogLevelTestRunnable.class.getSimpleName(), new LogLevelTestRunnable()).noLocalFiles()
+        .add(LogLevelTestSecondRunnable.class.getSimpleName(), new LogLevelTestSecondRunnable()).noLocalFiles()
+        .anyOrder()
+        .build();
+    }
+
+  }
+
+  @Test
+  public void testChangeLogLevel() throws Exception {
+    YarnTwillRunnerService runner = getTwillRunner();
+    runner.start();
+
+    // Set log level to DEBUG
+    TwillController controller = runner.prepare(new LogLevelTestApplication())
+      .setLogLevel(LogEntry.Level.DEBUG)
+      .addLogHandler(new PrinterLogHandler(new PrintWriter(System.out)))
+      .start();
+
+    // Lets wait until the service is running
+    final CountDownLatch running = new CountDownLatch(1);
+    controller.onRunning(new Runnable() {
+      @Override
+      public void run() {
+        running.countDown();
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+    Assert.assertTrue(running.await(200, TimeUnit.SECONDS));
+
+    // assert that log level is DEBUG
+    waitForLogLevel(controller, LogLevelTestRunnable.class.getSimpleName(),
```

Comment:
Actually I am not quite sure how to collect logs emitted by runnable in unit test. So in runnable I have checked the log level change: https://github.com/apache/twill/pull/14/files#diff-e7b3cb437b83c0f34231004947a7ff7eR61, which is simulated from the `LogLevelTestRun.class`, https://github.com/apache/twill/blob/master/twill-yarn/src/test/java/org/apache/twill/yarn/LogLevelTestRun.java#L60